File: src/main/java/EbbinghausReminder.java
import org.joda.time.DateTime;
import org.joda.time.Duration;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import static java.util.Arrays.asList;

public class EbbinghausReminder {
    private final DateTime today;
    private final DateTimeFormatter dateTimeFormatter = DateTimeFormat.forPattern("yyyy-MM-dd");
    private final List<Integer> daysBefore = asList(0, 1, 2, 4, 7, 15);

    public EbbinghausReminder(String today) {
        this.today = dateTimeFormatter.parseDateTime(today);
    }

    public List<String> datesToBeReviewed() {
        final List<String> dates = new ArrayList<>();
        for (Integer days : this.daysBefore) {
            dates.add(format(today.minusDays(days)));
        }
        return dates;
    }

    private String format(DateTime dateTime) {
        return dateTimeFormatter.print(dateTime);
    }

    public String randomDatesToBeReviewedAfter(String beginning) {
        DateTime end = DateTime.now().minusDays(daysBefore.get(daysBefore.size() - 1));
        Long days = new Duration(dateTimeFormatter.parseDateTime(beginning), end).getStandardDays();
        int randomDay = new Random().nextInt(days.intValue());
        return format(end.minusDays(randomDay));
    }
}

File: lib/ebbinghaus_reminder.rb
require 'date'

class EbbinghausReminder


  def initialize today_str
    @today = to_date today_str
  end

  def self.today
    EbbinghausReminder.new(DateTime.now.strftime('%Y-%m-%d'))
  end

  def self.today_is str
    EbbinghausReminder.new(str)
  end

  def dates_to_be_reviewed
    [0, 1, 2, 4, 7, 15].map { |days| @today - days }.map do |date|
      date.to_str
    end
  end

  def random_dates_to_be_reviewed_after start
    [(fifteen_days_before - Random.rand(1..fifteen_days_before - to_date(start))).to_str]
  end

  def fifteen_days_before
    @today - 15
  end

  def to_date(str)
    DateTime.strptime(str, '%Y-%m-%d')
  end

  private :to_date
end

class DateTime
  def to_str
    self.strftime('%Y-%m-%d')
  end
end